pallets/balances/src/lib.rs

```diff
@@ -1297,27 +1296,18 @@ where
         if amount.is_zero() || reasons.is_none() {
             return;
         }
-        let mut new_lock = Some(BalanceLock {
-            id,
-            amount,
-            reasons: reasons.into(),
-        });
-        let mut locks = Self::locks(who)
-            .into_iter()
-            .filter_map(|l| {
-                if l.id == id {
-                    new_lock.take().map(|nl| BalanceLock {
-                        id: l.id,
-                        amount: l.amount.max(nl.amount),
-                        reasons: l.reasons | nl.reasons,
-                    })
-                } else {
-                    Some(l)
-                }
-            })
-            .collect::<Vec<_>>();
-        if let Some(lock) = new_lock {
-            locks.push(lock)
+        let reasons = reasons.into();
+        let mut locks = Self::locks(who);
+        if let Some(pos) = locks.iter().position(|l| l.id == id) {
+            let slot = &mut locks[pos];
+            slot.amount = slot.amount.max(amount);
+            slot.reasons = slot.reasons | reasons;
+        } else {
+            locks.push(BalanceLock {
+                id,
+                amount,
+                reasons,
+            });
```

Comment:
My assumption it is a change for the complexity reduction only as it isn't affecting the code in any sense. 

If yes then either we keep as it as it is (In the favour of substrate upgrade) or please mention the reason of the change and make `Polymesh-note` so it will be easy to track the changes whenever we perform substrate upgrade

Reply:
The code is functionally equivalent, but reduces complexity in terms of:

- Performance:

  - The old code will allocate when doing `Self::locks(who)`, but then just iterates over that vector and creates another vector with `.collect<Vec<_>>()`. Moreover, the use of `.filter_map` will defeat any length-prediction, as there is no heuristic to guess what a random mapping function will do. 

  - The new code will do in-place modification of the vector returned by `Self::locks(who)`.

- Readability: The old code is a fairly roundabout way of doing "either modify or insert" whereas the new code is a more direct way of expressing this.

Ideally, the new code should be upstreamed to substrate.

Reply:
Yeah that I understand that's why I mention 

Please mention the reason of the change and make `Polymesh-note` (You can see these type of tags in the balances module) so it will be easy to track the changes whenever we perform substrate upgrade

Reply:
Done in https://github.com/PolymathNetwork/Polymesh/pull/521/commits/57d23114cac7fd0d14c35948f4ad89a73c0b2124.

Reply:
I see that you already did those changes